# Food Health Classifier

# “Mahsulot haqida ma’lumotlardan foydalanib, u qanchalik qayta ishlangan ekanini oldindan aytadigan model” qurish.

# Target Ustuni  = nova_group , nova_group 1–4 oralig‘ida 1: minimal qayta ishlangan (sog‘lom) , 4: yuqori qayta ishlangan (kamroq sog‘lom)

In [1]:
import pandas as pd
import logging

# Log fayl manzili
log_path = r"C:\Users\Rasulbek907\Desktop\Project_MP\Log\data_loader.log"

# Log sozlamalari
logging.basicConfig(
    filename=log_path,
    filemode='a',  # Append mode
    format='%(asctime)s - %(levelname)s - %(message)s',
    level=logging.INFO
)

try:
    logging.info("CSV fayl o'qilmoqda:...")
    df = pd.read_csv(r"C:\Users\Rasulbek907\Desktop\Project_MP\Data\Enginered_Data\sml_multiclass_dataset.csv")
    logging.info(f"Fayl muvaffaqiyatli o'qildi. Satırlar soni: {len(df)} ustunlar soni: {len(df.columns)}")
except Exception as e:
    logging.error(f"CSV faylni o'qishda xatolik: {e}")

In [2]:
df.head()

name_length  is_organic  is_sugar_free  quantity_value  \
0            2           0              0            33.0   
1            1           0              0           100.0   
2            2           0              0             2.0   
3            3           0              0             1.0   
4            1           0              0            33.0   

   quantity_unit_encoded  category_depth  country_count  product_age_days  \
0                      7               7              1              3671   
1                     14               1              2              3326   
2                     29               6              1              2590   
3                     29               7              1              4068   
4                      7               6              1              2365   

   created_year  created_month  main_category_encoded  nova_group  
0          2015              8                     53     3.28451  
1          2016              9                    459     3.00000  
2          2018              9                     53     1.00000  
3          2014              9                     53     1.00000  
4          2019              3                     68     3.28451

In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4993 entries, 0 to 4992
Data columns (total 12 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   name_length            4993 non-null   int64  
 1   is_organic             4993 non-null   int64  
 2   is_sugar_free          4993 non-null   int64  
 3   quantity_value         4993 non-null   float64
 4   quantity_unit_encoded  4993 non-null   int64  
 5   category_depth         4993 non-null   int64  
 6   country_count          4993 non-null   int64  
 7   product_age_days       4993 non-null   int64  
 8   created_year           4993 non-null   int64  
 9   created_month          4993 non-null   int64  
 10  main_category_encoded  4993 non-null   int64  
 11  nova_group             4993 non-null   float64
dtypes: float64(2), int64(10)
memory usage: 468.2 KB


# 🧾 Jadval ustunlari tavsifi

| №    | Ustun nomi                | Tarjima                              | Qisqacha ma’nosi                                                                                   |
| ---- | ------------------------- | ------------------------------------ | -------------------------------------------------------------------------------------------------- |
| 1️⃣  | **name_length**           | Nomi uzunligi                        | Mahsulot nomidagi belgilar soni (nom uzunligini bildiradi).                                        |
| 2️⃣  | **is_organic**            | Organik mahsulotmi                   | 1 — organik, 0 — oddiy mahsulot.                                                                   |
| 3️⃣  | **is_sugar_free**         | Shakarsizmi                          | 1 — shakarsiz, 0 — shakarlangan.                                                                   |
| 4️⃣  | **quantity_value**        | Miqdor qiymati                       | Mahsulot og‘irligi yoki hajmi (masalan, 100 ml, 250 g).                                            |
| 5️⃣  | **quantity_unit_encoded** | Miqdor birligi (kodlangan)           | Miqdor o‘lchov birligi (gramm, litr, dona va hokazo) raqam bilan kodlangan.                        |
| 6️⃣  | **category_depth**        | Kategoriya chuqurligi                | Mahsulot qaysi darajadagi kategoriya ichida joylashganini ko‘rsatadi (chuqurroq → aniqroq toifa).  |
| 7️⃣  | **country_count**         | Mamlakatlar soni                     | Mahsulot nechta mamlakatda mavjud yoki sotilayotganini bildiradi.                                  |
| 8️⃣  | **product_age_days**      | Mahsulot yoshi (kunlarda)            | Mahsulot bazaga kiritilgan kundan boshlab o‘tgan kunlar soni.                                      |
| 9️⃣  | **created_year**          | Yaratilgan yil                       | Mahsulot ma’lumotlari bazaga qo‘shilgan yil.                                                       |
| 🔟   | **created_month**         | Yaratilgan oy                        | Ma’lumot kiritilgan oy raqami (1–12).                                                              |
| 11️⃣ | **main_category_encoded** | Asosiy kategoriya (kodlangan)        | Mahsulotning asosiy turini ifodalovchi kod (masalan: ichimliklar, shirinliklar, sut mahsulotlari). |
| 12️⃣ | **nova_group**            | NOVA guruhi (qayta ishlash darajasi) | Mahsulot qayta ishlanganlik darajasi: <br> 1 — tabiiy, 4 — kuchli qayta ishlangan.                 |


In [4]:
df.describe()

name_length  is_organic  is_sugar_free  quantity_value  \
count   4993.00000  4993.00000         4993.0     4993.000000   
mean       3.47126     0.01462            0.0      294.821727   
std        2.12689     0.12004            0.0      230.069482   
min        1.00000     0.00000            0.0        0.000000   
25%        2.00000     0.00000            0.0      110.000000   
50%        3.00000     0.00000            0.0      250.000000   
75%        5.00000     0.00000            0.0      500.000000   
max       18.00000     1.00000            0.0     2041.165665   

       quantity_unit_encoded  category_depth  country_count  product_age_days  \
count            4993.000000     4993.000000    4993.000000       4993.000000   
mean               19.069497        7.237332       2.443421       2494.868816   
std                11.915155        3.399359       2.452280       1232.022279   
min                 1.000000        1.000000       1.000000          0.000000   
25%                14.000000        5.000000       1.000000       1543.000000   
50%                14.000000        7.000000       2.000000       2499.000000   
75%                14.000000       10.000000       3.000000       3452.000000   
max                71.000000       25.000000      25.000000       4991.000000   

       created_year  created_month  main_category_encoded   nova_group  
count   4993.000000    4993.000000            4993.000000  4993.000000  
mean    2018.152614       6.031444             262.622471     3.283936  
std        3.275715       3.332654             174.267565     0.974557  
min     2012.000000       1.000000               0.000000     1.000000  
25%     2016.000000       3.000000              75.000000     3.000000  
50%     2018.000000       6.000000             344.000000     4.000000  
75%     2021.000000       9.000000             415.000000     4.000000  
max     2025.000000      12.000000             519.000000     4.000000

# Asosiy ustunlar bo‘yicha statistik ko‘rsatkichlar

| Ustun nomi                | O‘rtacha qiymat (mean) | Eng kichik (min) | Eng katta (max) | Izoh                                                                                            |
| ------------------------- | ---------------------- | ---------------- | --------------- | ----------------------------------------------------------------------------------------------- |
| **name_length**           | 3.47                   | 1                | 18              | O‘rtacha nom uzunligi 3–4 so‘z atrofida, lekin ayrim mahsulot nomlari juda uzun (18 so‘zgacha). |
| **is_organic**            | 0.0146                 | 0                | 1               | Faqat ~1.5% mahsulotlar **organik**, bu juda kam.                                               |
| **is_sugar_free**         | 0.0                    | 0                | 0               | Hech bir mahsulot **shakarsiz** deb belgilanmagan (yoki ma’lumot yo‘q).                         |
| **quantity_value**        | 294.8                  | 0                | 2041            | O‘rtacha mahsulot miqdori 300 birlik atrofida, lekin juda katta farq mavjud.                    |
| **quantity_unit_encoded** | 19.06                  | 1                | 71              | Miqdor birliklari (ml, g, kg, L va hok.) keng diapazonda kodlangan.                             |
| **category_depth**        | 7.23                   | 1                | 25              | Kategoriyalar 1 dan 25 gacha chuqurlikda, ya’ni murakkab ierarxik tuzilma bor.                  |
| **country_count**         | 2.43                   | 1                | 25              | Mahsulotlar o‘rtacha 2–3 ta mamlakatda mavjud, ayrimlari 25 ta davlatda.                        |
| **product_age_days**      | 2494                   | 1                | 4991            | Mahsulotlar o‘rtacha 6–7 yillik (ehtimol, platformaga qo‘shilgan vaqtdan beri).                 |
| **created_year**          | 2018.15                | 2012             | 2025            | Ko‘pchilik mahsulotlar 2016–2021 oralig‘ida yaratilgan.                                         |
| **created_month**         | 6.03                   | 1                | 12              | Yil bo‘yicha teng taqsimlangan, o‘rtacha 6-oy (iyun) atrofida.                                  |
| **main_category_encoded** | 262.6                  | 0                | 519             | 500 ga yaqin asosiy kategoriya mavjud, turli xil mahsulot turlari.                              |
| **nova_group**            | 3.28                   | 1                | 4               | Ko‘pchilik mahsulotlar **3 yoki 4-darajali qayta ishlangan oziq-ovqat** hisoblanadi.            |


# Kuzatishlar va xulosalar

🥦 Organik mahsulotlar juda kam (1–2%), bu ma’lumotlar bazasida sog‘lom oziq-ovqatlar ulushi pastligini ko‘rsatadi.

🍬 Shakarsiz mahsulotlar yo‘q, bu ustun ehtimol to‘ldirilmagan yoki barchasi “False”.

⚖️ Miqdor (quantity_value) juda farq qiladi (0 dan 2000+ gacha), bu normalizatsiya yoki log-transformatsiya talab qiladi.

🌍 Country count ko‘rsatadi: ko‘pchilik mahsulotlar faqat 1–3 ta mamlakatda sotiladi, ayrimlari esa juda ko‘p joyda.

📅 created_year va product_age_days ustunlari vaqt bo‘yicha o‘zgarishlarni (trend) tahlil qilish uchun juda qulay.

🍔 nova_group qiymatlarining yuqoriligi shuni ko‘rsatadiki, ko‘pchilik mahsulotlar ko‘p qayta ishlangan (sog‘lom emas).

🗂️ main_category_encoded da 500 ga yaqin tur mavjud — bu multiclass muammo uchun yaxshi asos.

In [5]:
df["nova_group"].value_counts(normalize=True) * 100

nova_group
4.00000    52.553575
3.00000    22.010815
1.00000    11.355898
3.28451     9.833767
2.00000     4.245944
Name: proportion, dtype: float64

In [10]:
!pip install nbformat

  Using cached nbformat-5.10.4-py3-none-any.whl.metadata (3.6 kB)
  Using cached fastjsonschema-2.21.2-py3-none-any.whl.metadata (2.3 kB)
  Using cached jsonschema-4.25.1-py3-none-any.whl.metadata (7.6 kB)
  Using cached attrs-25.4.0-py3-none-any.whl.metadata (10 kB)
  Using cached jsonschema_specifications-2025.9.1-py3-none-any.whl.metadata (2.9 kB)
  Using cached referencing-0.37.0-py3-none-any.whl.metadata (2.8 kB)
Using cached nbformat-5.10.4-py3-none-any.whl (78 kB)
Using cached fastjsonschema-2.21.2-py3-none-any.whl (24 kB)
Using cached jsonschema-4.25.1-py3-none-any.whl (90 kB)
Using cached attrs-25.4.0-py3-none-any.whl (67 kB)
Using cached jsonschema_specifications-2025.9.1-py3-none-any.whl (18 kB)
Using cached referencing-0.37.0-py3-none-any.whl (26 kB)

   ---------------------------------------- 0/7 [fastjsonschema]
   ----------- ---------------------------- 2/7 [attrs]
   ----------------- ---------------------- 3/7 [referencing]
   ---------------------- -----------------


[notice] A new release of pip is available: 25.2 -> 25.3
[notice] To update, run: python.exe -m pip install --upgrade pip


In [11]:
import plotly.express as px
import pandas as pd

corr_matrix = df.corr()

fig = px.imshow(
    corr_matrix,
    text_auto=".2f",
    color_continuous_scale="RdBu_r",
    title="🔗 Korrelatsiya matritsasi (Plotly)",
    aspect="auto"
)

fig.update_layout(
    width=900,
    height=600,
    title_font=dict(size=22, color="black"),
    coloraxis_colorbar=dict(title="Korrelatsiya")
)

fig.show()


ValueError: Mime type rendering requires nbformat>=4.2.0 but it is not installed